Avoid non-requeue for MRs with determinstic external names.

nolancon opened the pull request:
<!--
Thank you for helping to improve Crossplane! Please read the contribution docs
(linked below) if this is your first Crossplane pull request.
-->

### Description of your changes
- Add a new ~~Annotation `crossplane.io/external-name-is-deterministic`~~ `WithDeterministicExternalName` reconciler option which allows the user to declare that a given resource is deterministically named by the external system.
- For MRs which are deterministically named, ignore the safeguard against leaked resources.

These changes enables the user to essentially "opt-out" of the current behaviour which blocks re-queuing of MRs in a scenario where creation is deemed incomplete. This safeguard is perfectly justified for MRs which are given non-deterministic names by the external system, preventing leaked resources (this scenario is described in detail [here](https://github.com/crossplane/crossplane/issues/3037#issuecomment-1110142427)). 

However, for resources which **are** deterministically named by the external system, the decision to not re-queue the MR is unnecessary as these resources are not in danger of leaking. For instance, [provider-ceph](https://github.com/linode/provider-ceph) creates S3 buckets by reconciling Bucket MRs - these s3 buckets are deterministically named and created idempotently. Therefore there is no danger in re-queuing after an initial "incomplete creation" failure (in fact it is expected). This kind of failure can (and does) happen regularly and in a system with 1000s of buckets leads to dozens of MRs which are not re-queued and therefore left in this pending state requiring manual intervention.

I have tested this patched version of the runtime against provider ceph. More details can be seen on this ~~[demo PR](https://github.com/linode/provider-ceph/pull/329)~~ [demo PR](https://github.com/linode/provider-ceph/pull/333) which also contains details on the existing workaround we use to attempt to mitigate the above scenario (but this is only best-effort and does not eradicate the problem completely).

Also, I'm happy to open a docs PR on this but would like to get feedback on this approach first, thanks.

<!--
Briefly describe what this pull request does, and how it is covered by tests.
Be proactive - direct your reviewers' attention to anything that needs special
consideration.

We love pull requests that fix an open issue. If yours does, use the below line
to indicate which issue it fixes, for example "Fixes #500".
-->

Fixes # 
https://github.com/crossplane/crossplane-runtime/issues/340 (maybe not fixes, but allows an "opt-out")

I have: <!--You MUST either [x] check or [ ] ~strike through~ every item.-->

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `earthly +reviewable` to ensure this PR is ready for review.
- [x] Added or updated unit tests.
- [ ] Linked a PR or a [docs tracking issue] to [document this change].
- [ ] Added `backport release-x.y` labels to auto-backport this PR.

Need help with this checklist? See the [cheat sheet].

[contribution process]: https://github.com/crossplane/crossplane/tree/main/contributing
[docs tracking issue]: https://github.com/crossplane/docs/issues/new
[document this change]: https://docs.crossplane.io/contribute/contribute
[cheat sheet]: https://github.com/crossplane/crossplane/tree/main/contributing#checklist-cheat-sheet

negz commented:
Thanks for taking a shot at this! I was thinking about something similar.

Could this be configured at the entire-type level though? I assume the if one `kind: Widget` has a deterministic external name, that'll be true for all `kind: Widget` resources, right? Ideally you wouldn't need to add this annotation to all of them.

Configuring this at the type level feels like the right way to go, but it's harder. It seems like a good use for the ManagedResourceDefinition type proposed in https://github.com/crossplane/crossplane/issues/2262.

negz commented:
Actually I guess this could just be handled at the code level via a new option to the MR reconciler. That'd mean provider authors would need to adopt it rather than being something end-users could specify though.

nolancon commented:
@negz I reverted the earlier changes and added a new commit with the ReconcilerOption approach to make the diff more readable. Here is an example PR for how this can be implemented by a downstream provider https://github.com/linode/provider-ceph/pull/333

negz commented on pkg/reconciler/managed/reconciler.go:
```diff
@@ -941,9 +955,10 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (resu
+	// an updated external name which is non-deterministic, we have leaked a
+	// resource. The safest thing to do is to refuse to proceed. However, if
+	// the resource has a deterministic external name, it is safe to proceed.
+	if meta.ExternalCreateIncomplete(managed) && !r.deterministicExternalName {
```
WDYT about debug logging this case (but not returning an error) even when there's a deterministic external name?

nolancon replied:
Sure, something like this https://github.com/crossplane/crossplane-runtime/pull/841/commits/6120cd37b63623eae817c75fc6efe893288621f6?

negz replied:
Yes that looks great.

negz reviewed (approved):
One suggestion but this LGTM. Thank you!
